### PR TITLE
support cos select

### LINKED ISSUE #4519: Presto support  COS select
We used Presto to test COS Select feature on csv files, but it‘s failed. 

We checked Presto source code and COS documentation, found that it was caused by the following reasons:
1. COS Select feature only support virtual-hosted style URL. But Presto only support path-style URL. The hard code is in PrestoS3ClientFactory class.
image

2. Cos Select feature don't support csv comments = "\uFDD0", but it's hard code in S3SelectCsvRecordReader class.
image

So, I think these parameters should be configurable in Presto. My patch fixed this problem.
[0001-support-cos-select.patch.txt](https://github.com/prestosql/presto/files/4953516/0001-support-cos-select.patch.txt)
 
Test fixed result:
(1) Fix Before, Presto's hive.properties is:
hive.s3.aws-access-key=AK
hive.s3.aws-secret-key=SK
hive.s3.endpoint=https://cos.ap-beijing.myqcloud.com
hive.s3select-pushdown.enabled=true
#hive.s3.path-style-access=false
#hive.s3.csv-comments=#

failed result
image

(2) Fix After, Presto's hive.properties is:
hive.s3.aws-access-key=AK
hive.s3.aws-secret-key=SK
hive.s3.endpoint=https://cos.ap-beijing.myqcloud.com
hive.s3select-pushdown.enabled=true
hive.s3.path-style-access=false
hive.s3.csv-comments=#

succeed result
image

COS Reference Documentation  https://intl.cloud.tencent.com/document/product/436/32472

### PR DESCRIPTION
**Pull request by Coderlxl**
Fixes #4519

### CODE REVIEW

File: presto-hive/src/main/java/io/prestosql/plugin/hive/s3/HiveS3Config.java [hunk 1]
```diff
@@ -66,6 +66,15 @@
...
     private boolean requesterPaysEnabled;
 
+    /*
+     * Sentinel unicode comment character (http://www.unicode.org/faq/private_use.html#nonchar_codes).
+     * It is expected that \uFDD0 sentinel comment character is not the first character in any row of user's CSV S3 object.
+     * The rows starting with \uFDD0 will be skipped by S3Select and will not be a part of the result set or aggregations.
+     * To process CSV objects that may contain \uFDD0 as first row character please disable S3SelectPushdown.
+     * TODO: Remove this proxy logic when S3Select API supports disabling of row level comments.
+     */
+    private String s3CsvComments = "\uFDD0";
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/s3/HiveS3Config.java:76]: Let's create a new config class `HiveS3SelectConfig` for this property. Name the config property `hive.s3select-pushdown.csv-comments` to match the other S3Select comments.

Later, we can move the existing S3Select configs from `HiveConfig` to the new `HiveS3SelectConfig`.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/s3select/S3SelectCsvRecordReader.java [hunk 2]
```diff
@@ -55,8 +47,11 @@ public S3SelectCsvRecordReader(
     }
 
     @Override
-    public SelectObjectContentRequest buildSelectObjectRequest(Properties schema, String query, Path path)
+    public SelectObjectContentRequest buildSelectObjectRequest(Properties schema, String query, Path path, Configuration configuration)
     {
+        HiveS3Config defaults = new HiveS3Config();
+        String s3CsvComments = configuration.get(S3_CSV_COMMENTS, defaults.getS3CsvComments());
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/s3select/S3SelectCsvRecordReader.java:53]: Since this CSV comments config is only used by S3Select, we can limit it to the S3Select code -- no need to pass it through Hadoop config:

1. Change `PrestoS3ClientFactory` to take the new `HiveS3SelectConfig` as mentioned in the other comment.
2. Store `csvComments` from the config in `PrestoS3ClientFactory`.
3. Add a method `getCsvComments()` to `PrestoS3ClientFactory`and use that in the `S3SelectCsvRecordReader` constructor to fetch the value.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/s3/PrestoS3FileSystem.java [hunk 3]
```diff
@@ -172,6 +172,7 @@
     public static final String S3_SKIP_GLACIER_OBJECTS = "presto.s3.skip-glacier-objects";
     public static final String S3_REQUESTER_PAYS_ENABLED = "presto.s3.requester-pays.enabled";
     public static final String S3_STORAGE_CLASS = "presto.s3.storage-class";
+    public static final String S3_CSV_COMMENTS = "presto.s3.csv-comments";
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/s3/PrestoS3FileSystem.java:175]: We can remove this after moving the config handling to `PrestoS3ClientFactory` (see other comment).

File: presto-hive/src/main/java/io/prestosql/plugin/hive/s3/HiveS3Config.java [hunk 4]
```diff
@@ -462,4 +471,17 @@ public HiveS3Config setRequesterPaysEnabled(boolean requesterPaysEnabled)
         this.requesterPaysEnabled = requesterPaysEnabled;
         return this;
     }
+
+    @NotNull
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/s3/HiveS3Config.java:475]: S3 ignores anything after the first character, so we also should enforce it is exactly one character with size validation (in addition to not null)
```java
@Size(min = 1, max = 1)
```

File: presto-hive/src/main/java/io/prestosql/plugin/hive/s3select/PrestoS3ClientFactory.java [hunk 5]
```diff
@@ -95,6 +96,7 @@ private AmazonS3 createS3Client(Configuration config)
         Duration connectTimeout = Duration.valueOf(config.get(S3_CONNECT_TIMEOUT, defaults.getS3ConnectTimeout().toString()));
         Duration socketTimeout = Duration.valueOf(config.get(S3_SOCKET_TIMEOUT, defaults.getS3SocketTimeout().toString()));
         int maxConnections = config.getInt(S3_SELECT_PUSHDOWN_MAX_CONNECTIONS, defaultMaxConnections);
+        boolean s3PathStyleAccess = config.getBoolean(S3_PATH_STYLE_ACCESS, defaults.isS3PathStyleAccess());
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/s3select/PrestoS3ClientFactory.java:99]: Can you make this a separate commit, since it's a bug fix, unrelated to the CSV comments change? Title it something like

> Respect path-style-access config for S3Select

**Review by electrum — COMMENTED** [review]
A few minor comments. Overall looks good

File: presto-hive/src/main/java/io/prestosql/plugin/hive/s3select/PrestoS3ClientFactory.java (hunk 5 above)
Comment (Coderlxl) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/s3select/PrestoS3ClientFactory.java:99]: Yes, it should be a separate commit

**Comment by tooptoop4** [comment]
still interested in this @Coderlxl ?

**Comment by colebow** [comment]
:wave: @Coderlxl - this PR is inactive and doesn't seem to be under development. If you'd like to continue work on this at any point in the future, feel free to re-open.
